openedx/features/journals/api.py

```diff
@@ -0,0 +1,276 @@
+            LOGGER.info("Unable to create DiscoveryApiClient because catalog integration not set up or enabled")
+            return None
+
+        try:
+            user = catalog_integration.get_service_user()
+        except ObjectDoesNotExist:
+            LOGGER.info("Unable to retrieve catalog integration service user")
+            return None
+
+        jwt = JwtBuilder(user).build_token([])
+        url = configuration_helpers.get_value('COURSE_CATALOG_API_URL', settings.COURSE_CATALOG_API_URL)
+        self.client = EdxRestApiClient(self.create_journals_url(url), jwt=jwt)
+
+    def create_journals_url(self, url):
+        """
+        rewrite the discovery url to point to journals endpoint
+        """
+        split_url = urlsplit(url)
+        override_url = urlunsplit((
+            split_url.scheme,
+            split_url.netloc,
+            JOURNALS_API_PATH,
+            split_url.query,
+            split_url.fragment,
+        ))
+        return override_url
+
+    def get_journals(self, orgs):
+        """
+        get_journals from discovery, filter on orgs is supplied
+        """
+        try:
+            if orgs:
+                response = self.client.journals.get(orgs=','.join(orgs), status='active')
+            else:
+                response = self.client.journals.get(status='active')
+            LOGGER.debug('response is type=%s', type(response))
+            return response.get('results')
+        except (HttpClientError, HttpServerError) as err:
+            LOGGER.exception(
+                'Failed to get journals from discovery-service [%s]',
+                err.content
+            )
+            return []
+
+    def get_journal_bundles(self, uuid=''):
+        """
+        get_journal_bundles from discovery on the base of uuid (optional)
+        """
+        try:
+            response = self.client.journal_bundles(uuid).get()
+        except (HttpClientError, HttpServerError) as err:
+            LOGGER.exception(
+                'Failed to get journal bundles from discovery-service [%s]',
+                err.content
+            )
+            return []
+        return response if uuid else response.get('results')
+
+
+class JournalsApiClient(object):
+    """
+    Class for interacting with the Journals Service
+    """
+    def __init__(self):
+        """
+        Initialize an authenticated Journals service API client by using the
+        provided user.
+        """
+        self.user = self.get_journals_worker()
+        jwt = JwtBuilder(self.user).build_token(['email', 'profile'], 16000)
+        self.client = EdxRestApiClient(
+            configuration_helpers.get_value('JOURNALS_API_URL', settings.JOURNALS_API_URL),
+            jwt=jwt
+        )
+
+    def get_journals_worker(self):
+        """ Return journals worker """
+        return User.objects.get(username=JOURNAL_WORKER_USERNAME)
+
+
+def fetch_journal_access(site, user):   # pylint: disable=unused-argument
```

Comment:
I am concerned about this access pattern.  Having LMS have a synchronous dependency on the catalog service isn't ideal.  However, it is cacheable and doesn't vary by user.

In this case, however, we are querying the Journals IDA for learner-specific data.  This cannot be cached across learners and is a synchronous call across IDAs.

A recommended alternative is to have the Journals IDA own the view for the Journals learner dashboard.  This way, integration between IDAs happen at the UI layer and not between backend services.  See [slides #23 to #28 of self-contained systems](https://speakerdeck.com/rstrangh/self-contained-systems-1?slide=23).  I'm happy to talk through this design pattern with the team.

For the time being, this may be fine as you test out this new feature.  However, we will run into scalability/reliability issues in the future.  We are feeling this pain now with other IDA integrations.